### PR TITLE
apiserver: fix leaks in logsink handler

### PR DESCRIPTION
**Pull request by axw**
While investigating https://bugs.launchpad.net/juju/+bug/1635311, I noticed a slow but steady increase in the number of goroutines in use by jujud. Taking a goroutine pprof snapshot, and then subsequently taking relative profiles, I found that there was >1 logsink handler. Since there's just one machine agent, this was unexpected.

A bug was introduced in 2.0 which causes the apiserver to not close out the websocket handler if there is a failure to read from the logsink client, except/until the apiserver.Server is stopped. If an error occurs when reading from the logsink client, the handler should close the connection.

Also, close the lumberjack WriterCloser when the apiserver stops.

### CODE REVIEW

File: apiserver/apiserver.go [hunk 1]
```diff
@@ -355,11 +364,13 @@ func (srv *Server) endpoints() []apihttp.Endpoint {
...
 
 	mainAPIHandler := srv.trackRequests(http.HandlerFunc(srv.apiHandler))
-	logSinkHandler := srv.trackRequests(newLogSinkHandler(httpCtxt, srv.logDir))
+	logSinkHandler := newLogSinkHandler(httpCtxt, srv.logSinkWriter)
 	logStreamHandler := srv.trackRequests(newLogStreamEndpointHandler(strictCtxt))
 	debugLogHandler := srv.trackRequests(newDebugLogDBHandler(httpCtxt))
 
-	add("/model/:modeluuid/logsink", logSinkHandler)
+	// TODO(axw) close the lumberjack writer.
+
```
Comment (anastasiamac) [inline apiserver/apiserver.go]: Is empty line needed?

File: apiserver/logsink.go [hunk 2]
```diff
@@ -23,23 +23,21 @@ import (
...
-
-	return &logSinkHandler{
-		ctxt: h,
-		fileLogger: &lumberjack.Logger{
-			Filename:   logPath,
-			MaxSize:    300, // MB
-			MaxBackups: 2,
-		},
+		// This isn't a fatal error so log and continue if priming fails.
+		logger.Warningf("Unable to prime %s (proceeding anyway): %v", logPath, err)
```
Comment (anastasiamac) [inline apiserver/logsink.go]: Warning or Info?

File: apiserver/apiserver.go (hunk 1 above)
Comment (axw) [inline apiserver/apiserver.go]: neither the empty line nor the TODO

File: apiserver/logsink.go (hunk 2 above)
Comment (axw) [inline apiserver/logsink.go]: Warning seems appropriate to me

**Comment by axw** [comment]
!!build!!

**Comment by axw** [comment]
$$merge$$

**Pull request merged**